rust/gitxetcore/src/environment/query_cache.rs

```diff
@@ -0,0 +1,215 @@
+    ///         query_cache.set_value(value.clone());
+    ///         value
+    ///      }
+    ///   };
+    ///
+    ///
+    pub fn new(
+        cache_dir: impl AsRef<Path>,
+        key_name: &str,
+        max_valid_seconds: u64,
+    ) -> Result<Self> {
+        let file_name = cache_dir.as_ref().join("query_cache").join(key_name);
+        Permission::current().create_dir_all(file_name.parent().unwrap())?;
+
+        Ok(Self {
+            query_value: None,
+            file_name,
+            max_valid_seconds,
+        })
+    }
+
+    /// Returns the value if it has been set or is cached and still valid.
+    pub fn get(&mut self) -> Option<String> {
+        if let Some(qv) = self.query_value.as_ref() {
+            return Some(qv.value.clone());
+        }
+
+        let Some(qv) = self.attempt_load_from_file() else {
+            return None;
+        };
+
+        if qv.query_age_in_seconds() < self.max_valid_seconds {
+            let ret = Some(qv.value.clone());
+            self.query_value = Some(qv);
+            ret
+        } else {
+            info!(
+                "Cached query age for {:?} too old, discarding.",
+                self.file_name
+            );
+            None
+        }
+    }
+
+    fn attempt_load_from_file(&self) -> Option<QueryValuesImpl> {
+        // if file doesn't exist return None
+
+        if !self.file_name.exists() {
+            debug!("{:?} does not exist; skipping load.", &self.file_name);
+            return None;
+        }
+
+        let Ok(file_contents) = std::fs::read_to_string(&self.file_name)
+            .debug_error(format!("Error reading version file {:?}", &self.file_name))
+        else {
+            return None;
+        };
+
+        let Ok(qv) = serde_json::from_str::<QueryValuesImpl>(&file_contents).info_error(format!(
+            "Error decoding version file contents from {:?}",
+            self.file_name
+        )) else {
+            debug!(
+                "Failed to parse query check information from {:?}.",
+                self.file_name
+            );
+            return None;
+        };
+
+        info!(
+            "Loaded cached query information {qv:?} from {:?}",
+            &self.file_name
+        );
+
+        Some(qv)
+    }
+
+    /// Verifies that the
+    pub fn set(&mut self, value: String) -> Result<()> {
+        let qv = QueryValuesImpl {
+            value,
+            query_time: Utc::now(),
+        };
+
+        // Now, save out the information here.
+        if let Ok(json_string) = serde_json::to_string_pretty(&qv).map_err(|e| {
+            debug!("Error serializing query values to JSON: {:?}", e);
+            e
+        }) {
+            debug!(
+                "CachedQueryInfo:save: Serializing cached query struct {qv:?} to {:?}",
+                &self.file_name
+            );
+
+            let _ = std::fs::write(&self.file_name, json_string).map_err(|e| {
```

Comment:
This is under `xet_home`, so ideally `Permission::create_file` before `std::fs::write`. As a side issue, in the future we may want to make `xet_home` a special object than PathBuf so it handles write with permission internally.

Reply:
It already creates the query_cache directory using Permission, so I'm not as worried about it here.